File: src/main/java/com/marsraver/WledFx/wled/WledNetworkScanner.java
package com.marsraver.WledFx.wled;

import com.marsraver.WledFx.wled.model.WledInfo;
import java.net.*;
import java.util.Collections;
import java.util.Enumeration;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class WledNetworkScanner {

    public static List<WledInfo> discover() throws Exception {
        String subnet = detectLocalSubnet();
        if (subnet == null) {
            System.err.println("❌ Could not determine local subnet.");
            return Collections.emptyList();
        }

        System.out.println("📡 Detected subnet: " + subnet + ".x");
        List<WledInfo> devices = new CopyOnWriteArrayList<>();
        ExecutorService pool = Executors.newFixedThreadPool(64);

        for (int i = 1; i < 255; i++) {
            final int host = i;
            pool.submit(() -> {
                String ip = subnet + "." + host;
                try {
                    WledClient client = new WledClient(ip);
                    WledInfo info = client.getInfo();
                    
                    devices.add(info);
                    System.out.println("✅ Found " + info);
                } catch (Exception ignored) {
                }
            });
        }

        pool.shutdown();
        pool.awaitTermination(25, TimeUnit.SECONDS);
        return devices;
    }

    private static String detectLocalSubnet() throws SocketException {
        Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
        while (interfaces.hasMoreElements()) {
            NetworkInterface iface = interfaces.nextElement();
            if (!iface.isUp() || iface.isLoopback() || iface.isVirtual()) continue;

            for (InterfaceAddress addr : iface.getInterfaceAddresses()) {
                InetAddress inetAddr = addr.getAddress();
                if (inetAddr instanceof Inet4Address && !inetAddr.isLoopbackAddress()) {
                    byte[] bytes = inetAddr.getAddress();
                    return String.format("%d.%d.%d",
                            bytes[0] & 0xFF,
                            bytes[1] & 0xFF,
                            bytes[2] & 0xFF);
                }
            }
        }
        return null;
    }

    public static void main(String[] args) throws Exception {
        System.out.println("🔍 Discovering WLED devices on local network...");
        List<WledInfo> devices = discover();
        System.out.printf("%n=== %d WLED devices found ===%n", devices.size());
        devices.forEach(System.out::println);
    }
}

File: src/main/kotlin/com/marsraver/wledfx/wled/WledNetworkScanner.kt
package com.marsraver.wledfx.wled

import com.marsraver.wledfx.wled.model.WledInfo
import java.net.Inet4Address
import java.net.InterfaceAddress
import java.net.NetworkInterface
import java.net.SocketException
import java.util.Collections
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

object WledNetworkScanner {

    @Throws(Exception::class)
    fun discover(): List<WledInfo> {
        val subnet = detectLocalSubnet()
        if (subnet == null) {
            System.err.println("❌ Could not determine local subnet.")
            return emptyList()
        }

        println("📡 Detected subnet: $subnet.x")
        val devices = CopyOnWriteArrayList<WledInfo>()
        val pool: ExecutorService = Executors.newFixedThreadPool(64)

        for (i in 1 until 255) {
            val host = i
            pool.submit {
                val ip = "$subnet.$host"
                try {
                    val client = WledClient(ip)
                    val info = client.getInfo()
                    info.ip = ip
                    devices.add(info)
                    println("✅ Found $info")
                } catch (_: Exception) {
                }
            }
        }

        pool.shutdown()
        pool.awaitTermination(25, TimeUnit.SECONDS)
        return Collections.unmodifiableList(devices)
    }

    @Throws(SocketException::class)
    private fun detectLocalSubnet(): String? {
        val interfaces = NetworkInterface.getNetworkInterfaces()
        while (interfaces.hasMoreElements()) {
            val iface = interfaces.nextElement()
            if (!iface.isUp || iface.isLoopback || iface.isVirtual) continue

            for (addr: InterfaceAddress in iface.interfaceAddresses) {
                val inetAddr = addr.address
                if (inetAddr is Inet4Address && !inetAddr.isLoopbackAddress) {
                    val bytes = inetAddr.address
                    return String.format(
                        "%d.%d.%d",
                        bytes[0].toInt() and 0xFF,
                        bytes[1].toInt() and 0xFF,
                        bytes[2].toInt() and 0xFF
                    )
                }
            }
        }
        return null
    }

    @JvmStatic
    fun main(args: Array<String>) {
        println("🔍 Discovering WLED devices on local network...")
        val devices = discover()
        println("\n=== ${devices.size} WLED devices found ===")
        devices.forEach { println(it) }
    }
}

